move softmax/logsoftmax to ATen

ngimel opened the pull request:
THCUNN kernels are mostly unchanged, with minimum changes to types so that more intermediate values are preserved in AccumT. 
Softmax/LogSoftmax from THNN are combined into a single templated function.
Remaining issues
 - in THNN accumulation for float used to be in double, ATen does not support this now, would need something like https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/AccumulateType.cuh, should we add it? With this PR, accumulation for float is in float. 
- in softmax_backward, self argument is added just so that gradient of self can be computed in double backward. Self tensor itself is not necessary neither for softmax_backward, nor for softmax_double_backward. I don't know if there's a better way of handling it. 
- scalar handling is somewhat awkward (I'm checking .dim() == 0, and do .view(1) for scalars), is there a better way of doing it?
- legacy tests are failing, because softmax is no longer legacy. I can rewrite legacy softmax forward to use torch.softmax instead of [backend].SoftMax_UpdateOutput, but what to do with backward? Softmax backward is no longer exposed.

Commit:
move softmax/logsoftmax to ATen

bddppq commented:
@onnxbot retest this please

vadimkantorov commented:
a naming question: we have `logsigmoid` and `log_softmax`. If `log_softmax` is being moved to ATen, should it be renamed to match `logsigmoid`? or left as is

ngimel commented:
@vadimkantorov good point, I can rename if there's an agreement from core devs.

ezyang commented:
CC @colesbury on the legacy test issues. Last time I chatted with folks about it, we weren't planning to delete the legacy support, so something might have to be done.

ezyang commented:
> in THNN accumulation for float used to be in double, ATen does not support this now, would need something like https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cuda/AccumulateType.cuh, should we add it? With this PR, accumulation for float is in float.

Yes, that would be great. I'd probably make another version of this template for "AccAccumulateType" or similar (following the TH convention)

> in softmax_backward, self argument is added just so that gradient of self can be computed in double backward. Self tensor itself is not necessary neither for softmax_backward, nor for softmax_double_backward.

I'm a little perplexed by the double-backwards situation; if `_backward` has no data dependence on `self`, then there shouldn't be any gradient formula for the double backwards, right?

CC @gchanan about scalars

ngimel commented:
1) For legacy, I'd need to figure out how to expose softmax_backward to python. Any advice?
2) I'll add cpu accumulate type template that follows what TH/THNN has now, and use it on the cpu paths (as a side note, I see cpu reductions now being moved from TH to ATen, they should face the same issue).
3) It has an implicit dependence on `self` via `output/result` that is actually used in the gradient formula (a side note: in derivatives.yaml, for the first derivative softmax_backward has to be declared with `result` argument, not `output`, otherwise generator does not properly insert unpacking code. For the second derivative, `softmax_backward` has to be declared with `output` argument, otherwise compiler complains about something, don't remember what).
So in this piece of code
```
name: softmax_backward(Tensor grad_output, Tensor output, int64_t dim, Tensor self)
grad_output: softmax_backward(grad, output, dim, self)
self: softmax_double_backward(grad, grad_output, dim, output)
```
`self` in the first line is just so that there can be `self` in the third line. Does it make sense?

apaszke commented:
Re 3: that makes sense, but IIRC that's not what we used to do some time ago. Isn't it possible to rewrite the derivative of `grad_output` in terms of `output`?

ezyang commented:
@apaszke No, the issue is different; none of the gradient formulas reference self.

I think we should leave it for now but eventually we should figure out what exactly is going on here. I would have thought that maybe autograd should have worked out that if there's a differential on output, then you need to in turn compute the differential on self because output was computed from self.

apaszke commented:
If they don't reference `self` then why does it need to be passed around at all? I think we should resolve this before we merge this. Softmax is often used in places that are pretty critical for memory usage, and should be efficient.

ngimel commented:
It's passed around so that double backward on self can be defined, in the 3-line snippet I posted above. It's no worse than it is now (THNN/THCUNN also does not use input in backward, yet it is there as an argument)
https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L761
https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml#L962
but I'm all for resolving it.

apaszke commented:
Hmm I see now. Alright, that sounds a bit complicated, so we don't have to block on that.

Commit:
specify cpu and gpu accum types

Commit:
use accreal for CPU

Commit:
expose softmax backward to python, fix legacy interface

Commit:
fix Distributions.cu to use common AccumulateType

Commit:
fix cuda 8 build

ngimel commented:
Rebased on master, fixed legacy by calling `*backward`  `*backward_data` (native functions ending in `backward` are blacklisted from being exposed in python). Don't particularly like this solution, but can't find any other way. 
Moved AccumulateType to be cpu/gpu accumulate type, with accum types set to what they are now in TH/THC. For cpu float softmax, behavior now replicates THNN, accumulation is in double. Had to change includes in a few  existing files to `ATen/AccumulateType.h` instead of `ATen/cuda/AccumulateType.cuh`, and fix the source.

vadimkantorov commented:
@apaszke what's your view on making log_softmax and logsigmoid naming consistent? https://github.com/pytorch/pytorch/pull/6786#issuecomment-383037489

apaszke commented:
If there's something to be renamed it should be `logsigmoid`. `log_softmax` is used in half of PyTorch scripts and that's too much to change. I also like the separated syntax more.

apaszke commented on torch/nn/functional.py:
```diff
@@ -943,7 +945,8 @@ def log_softmax(input, dim=None, _stacklevel=3):
     if dim is None:
         dim = _get_softmax_dim('log_softmax', input.dim(), _stacklevel)
-    return torch._C._nn.log_softmax(input, dim)
+#    return torch._C._nn.log_softmax(input, dim)
```
Can you please remove those comments?

apaszke commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -125,49 +197,51 @@ __global__ void cunn_SpatialSoftMaxForward(
+ 
       if (blockDim.x > 1) {
-        T max_input = THCNumerics<T>::min();
+        accscalar_t max_input = THCNumerics<accscalar_t>::min();
```
Why do we keep max in accscalar now? That's not how it used to be

ngimel replied:
1) comparison operators for half work on floats only anyway, so there will be conversions. Might as well convert in the beginning and keep intermediates in float
2) reduced max will still have to be converted to accscalar in epilogue, either when epilogue is constructed or for every call, so there's also nothing to be gained by keeping it in scalar

apaszke commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -125,49 +197,51 @@ __global__ void cunn_SpatialSoftMaxForward(
-        for (uint32_t d = 0; d < dim_size; d++) {
-          const T value = input[data_offset + d * dim_stride];
-          max_input = THCNumerics<T>::ge(max_input, value) ? max_input : value;
+        accscalar_t max_input = THCNumerics<accscalar_t>::min();
```
Same here

apaszke reviewed (commented):
Not a complete review by any means, but looks ok at a glance. Just curious why some `T` changed to `AccumT` in the kernels

Commit:
delete commented out lines

SsnL commented:
@ngimel can you kindly rebase please?

Commit:
rebase on master, fix breakages

ngimel commented:
I don't see an actual error on failed trusty build, other than gcc exited with error code 1 - did it time out? For failed xenial-cuda, the failure also does not look related (there's no softmax in the test script).

ngimel commented:
Hmm, cuda9-cudnn7 failed again, how do I run test/cpp/ tests locally?

ngimel commented:
Interestingly, I can make adagrad fail on master by commenting out integration.cpp from test_api sources (because I was too lazy to wait for mnist training), in this case adagrad goes for 3000 epochs with loss stuck at 0.48. 
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test_api is a Catch v2.2.1 host application.
Run with -? for options

-------------------------------------------------------------------------------
optim
  adagrad
-------------------------------------------------------------------------------
/raid/pytorch/test/cpp/api/optim.cpp:56
...............................................................................

/raid/pytorch/test/cpp/api/optim.cpp:64: FAILED:
```
Optimizer tests seem kind of flaky

ngimel commented:
So given that the failing test is flaky (#7288), and this PR actually does not touch any of the things the failing test is testing this should be good to go?

cpuhrsch commented:
Hey @ngimel, are you planning on further changing the CPU implementation? I'm currently working on vectorizing softmax and I have just finished what you merged on my end as well haha.

ngimel commented:
No, I did not plan on further optimizing CPU path, but now that it's in ATen it should be easier for you to work on.

cpuhrsch commented:
Thanks @ngimel!
